File: Python/BasicDataStructures/Queue.py
class Queue:
    def __init__(self, capacity):
        self.capacity = capacity
        self.front = self.size = 0
        self.rear = capacity-1
        self.q = [None]*capacity
    def isFull(self): return self.size == self.capacity
    def isEmpty(self): return self.size == 0
    def addQ(self, x):
        if self.isFull():
            print("Full")
            return
        self.rear = (self.rear+1)%(self.capacity)
        self.q[self.rear]=x
        self.size+=1
        print(f"{x} is enqueued to queue")
    def delQ(self):
        if self.isEmpty():
            print("Empty")
            return
        print(f"{self.q[self.front]} is dequeued from queue")
        self.front = (self.front+1)%(self.capacity)
        self.size -=1
    def getFront(self):
        if self.isEmpty():
            print("Queue is empty")
        print(f"Rear item is: {self.q[self.rear]}")
    def getRear(self):
        if self.isEmpty():
            print("Queue is empty")
        print(f"Front item is: {self.q[self.front]}")      
queue = Queue(30)
queue.addQ(10)
queue.addQ(20)
queue.addQ(30)
queue.addQ(40)
queue.delQ()
queue.getFront()
queue.getRear()

File: C++/DataStructures/Queue/Queue.cpp

#include <iostream>
using namespace std;

class Queue {
public:
    int front, rear, size;
    unsigned capacity;
    int* array;
};

Queue* createQueue(unsigned capacity) {
    Queue* queue = new Queue();
    queue->capacity = capacity;
    queue->front = queue->size = 0;

    queue->rear = capacity - 1;
    queue->array = new int[queue->capacity];
    return queue;
}

int search(Queue* queue, int key) {
    for (int i = 0; i <= queue->size;i++) {
        if (queue->array[i] == key)return i - queue->front;
    }
    return -1;
}

bool isFull(Queue* queue) {
    return(queue->size == queue->capacity);
}

bool isEmpty(Queue* queue) {
    return(queue->size == 0);
}

void enqueue(Queue* queue, int item) {
    if (isFull(queue))return;
    queue->rear = (queue->rear + 1) % queue->capacity;
    queue->array[queue->rear] = item;
    queue->size = queue->size + 1;
    cout << item << " enqueued to queue\n";
}

int dequeue(Queue* queue) {
    if (isEmpty(queue))return INT_MIN;
    int item = queue->array[queue->front];
    queue->front = (queue->front + 1) % queue->capacity;
    queue->size = queue->size - 1;
    return item;
}

int front(Queue* queue) {
    if (isEmpty(queue))return INT_MIN;
    return queue->array[queue->front];
}

int rear(Queue* queue) {
    if (isEmpty(queue))
        return INT_MIN;
    return queue->array[queue->rear];
}
int main()
{
    Queue* queue = createQueue(1000);

    enqueue(queue, 10);
    enqueue(queue, 20);
    enqueue(queue, 30);
    enqueue(queue, 40);

    cout << dequeue(queue) << " dequeued from queue\n";

    cout << "Front item is " << front(queue) << endl;

    cout << "Rear item is " << rear(queue) << endl;

    cout << "Is 40 in queue? ";
    int index = search(queue, 40);
    index != -1 ? cout << "Yes,  at a distance of " << index << " from front element.\n" : cout << "No."<<endl;
    index = search(queue, 10);
    cout << "Is 10 in queue? ";
    index != -1 ? cout << "Yes,  at a distance of " << index << " from front element." : cout << "No.";
    return 0;
}

